### PR TITLE
Add GPT-J support (#285)

### PR DESCRIPTION
**Pull request by lerouxrgd**
Remaining points:
- [ ] The `GptJModelResources` rust_model.ot files should be added to HF Hub (I have never tried to add models there but I can try if you want)
  - [x] PR for [gpt-j-tiny-random](https://huggingface.co/anton-l/gpt-j-tiny-random/discussions/1)
  - [ ]  PR for [gpt-j-6b float32](https://huggingface.co/EleutherAI/gpt-j-6B/discussions/5)
  - [ ]  PR for gpt-j-6b float16
- [x] `GptJConfig::use_float16` defaults to true, do you want to keep it this way ?
- [x] `GptJConfig::preload_on_cpu` defaults to true, do you want to keep it this way ?

Note that `GptJConfig::preload_on_cpu` makes it possible to load `GPT-J 6B` (both float16 and float32 with conversion to float16 on the fly) on a T4 GPU (with ~16Gb VRAM). Without this flag there are memory issues when loading the weights with `var_store.load(weights_path)`.

### CODE REVIEW

File: src/gpt_j/attention.rs
Comment (guillaume-be) [inline src/gpt_j/attention.rs:302]: Interesting first time I come across this

File: src/gpt_j/gpt_j_model.rs
Comment (guillaume-be) [inline src/gpt_j/gpt_j_model.rs:305]: You probably want to wrap this into a `LayerState`

File: src/gpt_j/gpt_j_model.rs
Comment (guillaume-be) [inline src/gpt_j/gpt_j_model.rs:351]: Since this method returns a `Result` you could raise an `Error` rather than panicking

File: src/gpt_j/gpt_j_model.rs
Comment (guillaume-be) [inline src/gpt_j/gpt_j_model.rs:372]: Is this valid for both `fp32` and `fp16`?

File: tests/gpt_j.rs
Comment (guillaume-be) [inline tests/gpt_j.rs:1]: Can you please add this test to the integration tests as well (after https://github.com/guillaume-be/rust-bert/blob/78da0f4814da00139551b615d6d06e7e92f59a79/.github/workflows/continuous-integration.yml#L117)

File: tests/gpt_j.rs
Comment (guillaume-be) [inline tests/gpt_j.rs:43]: Thank you for adding tests. 

I would suggest the following:
- For the `tiny` test that will run in the pipeline, please instead do a forward pass through the model by passing an input tensor and a past tensor, and compare that to the output generated by Python (e.g. to the 4th decimal, see https://github.com/guillaume-be/rust-bert/blob/78da0f4814da00139551b615d6d06e7e92f59a79/tests/m2m100.rs#L64 for example)
- It would be nice to add a test for a full-sized model. Can you add a test that generates a sensible output to test the generation process using float16, on CPU and flagging the test as `#[cfg_attr(not(feature = "all-tests"), ignore)]` )?

File: src/gpt_j/gpt_j_model.rs
Comment (guillaume-be) [inline src/gpt_j/gpt_j_model.rs:55]: It would be good to clarify the license for this model with the author or the Huggingface team

**Review by guillaume-be — COMMENTED** [review]
Hello @lerouxrgd 
This is great - thank you very much for this PR! Could you please open pull requests on the Huggingface model hub to upload the model weights that are registered in the crate (you may want to refer this PR as part of the PR there)? It would also be great to extend the test with a valid generation example - the crate has a flag for disabling large tests by default (in the CI pipeline), but allows running them manually on a larger machine.

File: src/gpt_j/mod.rs
Comment (lerouxrgd) [reply src/gpt_j/mod.rs:49]: Indeed it's not, needed I've taken it from GPT2. I'll remove the `pub(crate)`.

File: src/gpt_j/attention.rs
Comment (lerouxrgd) [reply src/gpt_j/attention.rs:90]: Good  point, I originally used a Tensor as in the Python implementation. However it looks like the broadcasting is only available for`f64` with `<Scalar as From<f64>>`, which is fine I guess.

File: src/gpt_j/attention.rs
Comment (lerouxrgd) [reply src/gpt_j/attention.rs:113]: Right, but I think that converting to half as soon as possible is safer, especially since at layer 28 the memory usage starts to be important.

File: src/gpt_j/attention.rs
Comment (lerouxrgd) [reply src/gpt_j/attention.rs:302]: I wasn't 100% confident it would work when I first wrote it, but it works !

File: src/gpt_j/gpt_j_model.rs
Comment (lerouxrgd) [reply src/gpt_j/gpt_j_model.rs:372]: It looked OK in my tests (same text generated as in Python) with models of both precisions. Do  you see another kind of test for this in particular ?

File: tests/gpt_j.rs
Comment (lerouxrgd) [reply tests/gpt_j.rs:43]: I'll work on the test you suggested for the `tiny` model.

As for the full-sized model using float16 it won't run on the CPU, when I tried I got a Torch error: "LayerNormKernelImpl" not implemented for 'Half'. I think some cuda kernels are not available for inference on the CPU when using float16. Using the folat32 version requires at least 32 Gb of RAM and I'm not sure the CI has that much available. What do you think ?

File: utils/convert_model.py
Comment (lerouxrgd) [reply utils/convert_model.py:34]: done

File: src/gpt_j/mod.rs
Comment (lerouxrgd) [reply src/gpt_j/mod.rs]: Done

File: src/gpt_j/attention.rs
Comment (lerouxrgd) [reply src/gpt_j/attention.rs]: Done

File: src/gpt_j/attention.rs
Comment (lerouxrgd) [reply src/gpt_j/attention.rs]: Done

File: src/gpt_j/gpt_j_model.rs
Comment (lerouxrgd) [reply src/gpt_j/gpt_j_model.rs]: Done

File: src/gpt_j/gpt_j_model.rs
Comment (lerouxrgd) [reply src/gpt_j/gpt_j_model.rs]: Done

File: src/gpt_j/gpt_j_model.rs
Comment (lerouxrgd) [reply src/gpt_j/gpt_j_model.rs:383]: I guess the test you suggested would the `tiny` model will help, I'll let  you know the results.

File: src/gpt_j/gpt_j_model.rs
Comment (guillaume-be) [inline src/gpt_j/gpt_j_model.rs:301]:
```suggestion
    /// let mut past: Vec<Option<LayerState>> = Vec::with_capacity(config.n_layer as usize);
```

File: src/gpt_j/gpt_j_model.rs
Comment (guillaume-be) [inline src/gpt_j/gpt_j_model.rs:303]:
```suggestion
    ///     past.push(Some(LayerState {
```

File: src/gpt_j/gpt_j_model.rs
Comment (guillaume-be) [inline src/gpt_j/gpt_j_model.rs:322]:
```suggestion
    ///     }))
```

File: src/gpt_j/gpt_j_model.rs
Comment (guillaume-be) [inline src/gpt_j/gpt_j_model.rs:366]:
```suggestion
                    let length = value.len();
                    (value, length )
```

File: tests/gpt_j.rs
Comment (lerouxrgd) [reply tests/gpt_j.rs:1]: Done

**Comment by guillaume-be** [comment]
@lerouxrgd to my understanding this pull request is held up by the pending model weights pull request on the huggingface hub. It seems that hub model contributors do not automatically get notifications when an issue/PR is created on the models, you may want to try reaching out directly to the authors (if you find their contact information) or the Huggingface maintainers?

**Comment by lerouxrgd** [comment]
Yes, I have reached out to maintainers but I don't have a final answer yet.

**Comment by guillaume-be** [comment]
Hello @lerouxrgd ,

If it is difficult merging the Rust weights, I suggest updating this pull request removing the pretrained weights and replacing it with a comment indicating how to generate the set of weight from the Python version. If and when the original model authors approve hosting Rust weights (or if they encourage creating a new model on the Hub for the Rust weights), you could open a new pull request making these weights available to the community.

What do you think?

**Comment by lerouxrgd** [comment]
Hello @guillaume-be ,

It's sad to see that the model's maintainer aren't really willing to add the Rust weights nor giving constructing feedback. Therefore I think you're right we can explain in comments how to create them. I have updated the PR accordingly, let me know what you think.

File: src/gpt_j/gpt_j_model.rs
Comment (guillaume-be) [reply src/gpt_j/gpt_j_model.rs:390]: I am not sure if you had the chance to test this - Could you please run the Python version was half precision weights and update the "correctness" integration tests with this the transformation?

File: tests/gpt_j.rs
Comment (guillaume-be) [inline tests/gpt_j.rs:56]: If these weights are not available could we please remove this from integration tests? This could be added as an example instead if you wish to be run manually.

You could also try loading the `pytorch_model.bin` model file with the ltest `tch-rs` [update](https://github.com/LaurentMazare/tch-rs/commit/75074f3253d9cc4d5f2988fc03ad16b6101b90d5). It does not work for all model files yet, but it compatible with GPT-J files this would allow readily loading the model from pretrained resources.

Comment (lerouxrgd) [reply tests/gpt_j.rs:56]: It looks like I also get a "Expected GenericDict but got None" error as mentioned in [tch-rs/issues/595](https://github.com/LaurentMazare/tch-rs/issues/595#issuecomment-1364108766)

File: src/gpt_j/gpt_j_model.rs
Comment (lerouxrgd) [reply src/gpt_j/gpt_j_model.rs:390]: Well using half precision doesn't work well on CPU (as they usually don't support float16), so this would require Cuda and GPU to run, which is not what the GitHub CI provides I guess.

I will still try to run it on a GPU on my side to check the correctness.

Comment (guillaume-be) [reply src/gpt_j/gpt_j_model.rs:390]: I usually try to run a broader set of test using a GPU - maybe a solution could be to add a test for half precision that runs only on GPU? i.e.
```rust
#[test]
fn test_gpt_j_half_precision {
    if Cuda::is_available {
       [... tests here ...]
    }
}
```

File: tests/gpt_j.rs
Comment (guillaume-be) [reply tests/gpt_j.rs:56]: Ah that's unfortunate, thank you for checking anyway. It may make sense to move this code snippet to the examples if the download/conversion cannot be easily automated - what do you think?

File: src/gpt_j/gpt_j_model.rs
Comment (lerouxrgd) [reply src/gpt_j/gpt_j_model.rs:390]: While giving it a try I realized that with latest Python `transformers` I don't get the same results for the "correctness" test anymore. After investigations it looks like in `4.21.0` HF [made changes](https://github.com/huggingface/transformers/commit/360719a6a4e86c99cb852e844d30712eabb04baf) to the implementation of GPT-J. I will try to adapt the PR accordingly first.

Comment (lerouxrgd) [reply src/gpt_j/gpt_j_model.rs:390]: Interestingly the generation test is still fine ...

**Comment by guillaume-be** [comment]
Hello @lerouxrgd ,
I see you opened https://github.com/LaurentMazare/tch-rs/issues/604 .

I want to point you out to the existing utilities in that can be found in https://github.com/guillaume-be/rust-bert/blob/master/src/common/kind.rs may be useful if related to completing this pull request.

Please feel free to extend if you need values that are not `MAX/Infinity`

**Comment by lerouxrgd** [comment]
Oh I  wasn't aware of it, thanks for pointing it out.

Although I think it's just half of the story, the other half being the `impl From<half::f16> for Scalar` etc for all `Element`s (or `Kind` ?) to be able to multiply them with a `Tensor`. Let's see how this issue goes in `tch-rs`.

**Comment by guillaume-be** [comment]
Could you please share a pointer what you are trying to implement?

**Comment by lerouxrgd** [comment]
So [this line](https://github.com/lerouxrgd/rust-bert/blob/gpt-j/src/gpt_j/gpt_j_model.rs#L390):

```rs
let attention_mask: Tensor = (1.0 - attention_mask) * (-10000.0);
```

works because of [impl From< f64> for Scalar](https://docs.rs/tch/0.10.1/tch/struct.Scalar.html#impl-From%3Cf64%3E-for-Scalar) and [impl< S: Into< Scalar>> Mul< S> for Tensor](https://docs.rs/tch/0.10.1/tch/struct.Tensor.html#impl-Mul%3CS%3E-for-Tensor), however as mentioned in the `tch-rs` issue, attention mask in HuggingFace Python code is now implemented  as follows since version `4.21.0`:

```python
attention_mask = (1.0 - attention_mask) * torch.finfo(self.dtype).min
```

So we need 2 things to port this code correctly:

1. Something in Rust similar to `torch.finfo(self.dtype).min` to get the correct value (instead of hardcoded -10000.0)
2. The ability to multiply a `Tensor` by the resulting min value. And in case of `float16` this is not doable now as I mentioned in the `tch-rs` issue

**Comment by lerouxrgd** [comment]
Writing this I realize that even the `1.0 - attention_mask` part is probably not correct as the same remark holds for `impl< S: Into< Scalar>> Add< S> for Tensor`, and some implicit conversion to `f64` is already happening ... Which might defeat the point of using `float16` I guess...

So maybe simply using your function for getting `Kind`'s min is enough. What do you think ?

**Comment by guillaume-be** [comment]
The attention mask may be of type `Kind::Bool` or `Kind::Int64` when the subtraction happens. I was working on a `LongT5` implementation this morning and implemented that same mechanism as:
```
(attention_mask.ones_like()
            - attention_mask.to_kind(input_embeddings.kind()))
            * get_negative_infinity(input_embeddings.kind()).unwrap()
```
where `input_embedding` is the input tensor to this layer (you have have `hidden_states` or other you can use as a reference).
I have not yet tested it, but it compiles fine - maybe this helps?

**Comment by lerouxrgd** [comment]
I see, thanks I'll give it a try on Monday !

**Comment by lerouxrgd** [comment]
So it compiles fine, but sadly it doesn't change the results... There is still some changes made in HF between 4.20.1 and 4.21.0 that affect the correctness results and I haven't found what they are.

**Comment by guillaume-be** [comment]
@lerouxrgd I noticed a discrepancy when implementing `longt5`: the masked position are not filled with `NEGATIVE_INFINITY` but with `MIN` . I have created new method to access either the `infinity` or `min/max` values for the different `dtypes`: can you please try using the `get_min` method:
https://github.com/guillaume-be/rust-bert/blob/d7e9c036940aefebe81c083841e4ff3bad36bd56/src/common/kind.rs#L42
that is merged as part of https://github.com/guillaume-be/rust-bert/pull/333 and see if this helps?

**Comment by lerouxrgd** [comment]
Yes this what I tried already as you can see [here](https://github.com/guillaume-be/rust-bert/pull/288/files#diff-034db0260f0646772831690ccd33fc3ea073890310a4981dd10575f7addbd7a1R44-R61) and [here](https://github.com/guillaume-be/rust-bert/pull/288/files#diff-8d488f786d2b62317fca13711b2e9a46445e5d228efe441854be6fbd09238ba0R198)

**Comment by lerouxrgd** [comment]
@guillaume-be Good news, after a long println debugging session I've finally figured out that the main issue was that I didn't set the attention mask in my unit test. I guess this was still OK in the previous attention version but not with the latest one.

File: tests/gpt_j.rs
Comment (guillaume-be) [reply tests/gpt_j.rs:56]: @lerouxrgd it looks like the PR is almost ready to be merged -- any chance this could be moved to an example? This would give more visibility to this new feature, and keep all tests independent of pre-download/conversion steps

**Comment by guillaume-be** [comment]
> @guillaume-be Good news, after a long println debugging session I've finally figured out that the main issue was that I didn't set the attention mask in my unit test. I guess this was still OK in the previous attention version but not with the latest one.

This is awesome, thank you so much for your patience for getting to the bottom of this! I have one last request (move of the run with GPT-J actual weights to examples) and this should be good to merge

File: tests/gpt_j.rs
Comment (lerouxrgd) [reply tests/gpt_j.rs]: That's exactly what I was doing :)

Comment (lerouxrgd) [reply tests/gpt_j.rs]: Done (And I have re-checked that it works fine on a T4 GPU)

**Comment by lerouxrgd** [comment]
Ok I think everything is fine now. I have made the correctness test runnable on both CPU and GPU (forcing it to use float16 in this case, which in turn lowers precision to 1e-2 instead of 1e-4).

Let me know if some other changes are needed.

**Review by guillaume-be — APPROVED** [review]
This is great - thanks again for this PR @lerouxrgd !

**Pull request merged**
